include/nbl/video/utilities/IGPUObjectFromAssetConverter.h

```diff
@@ -866,103 +866,35 @@ auto IGPUObjectFromAssetConverter::create(const asset::ICPUImage** const _begin,
-            barrier.barrier.dstAccessMask = asset::EAF_TRANSFER_WRITE_BIT;
-
-            IGPUCommandBuffer::SImageMemoryBarrier toTransferDst = {};
-            toTransferDst.barrier.srcAccessMask = static_cast<asset::E_ACCESS_FLAGS>(0u);
-            toTransferDst.barrier.dstAccessMask = asset::EAF_TRANSFER_WRITE_BIT;
-            toTransferDst.oldLayout = asset::IImage::EL_UNDEFINED;
-            toTransferDst.newLayout = asset::IImage::EL_TRANSFER_DST_OPTIMAL;
-            toTransferDst.srcQueueFamilyIndex = transferFamIx;
-            toTransferDst.dstQueueFamilyIndex = transferFamIx;
-            toTransferDst.image = core::smart_refctd_ptr<video::IGPUImage>(img);
-            toTransferDst.subresourceRange.aspectMask = asset::IImage::EAF_COLOR_BIT; // this probably shoudn't be hardcoded
-            toTransferDst.subresourceRange.baseMipLevel = 0u;
-            toTransferDst.subresourceRange.levelCount = img->getCreationParameters().mipLevels;
-            toTransferDst.subresourceRange.baseArrayLayer = 0u;
-            toTransferDst.subresourceRange.layerCount = cpuimg->getCreationParameters().arrayLayers;
-
-            cmdbuf_transfer->pipelineBarrier(
-                asset::EPSF_TRANSFER_BIT,
-                asset::EPSF_TRANSFER_BIT,
-                asset::EDF_NONE,
-                0u, nullptr,
-                1u, &barrier,
-                1u, &toTransferDst);
-
-            // Note: cpuimg->getRegions() doesn't give correct values for this because they
-            // get set by the asset loader which doesn't have enough information about what
-            // kind of subresources will be created for this resource.
-            // For this function we want the entirety of ICPUImage converted to IGPUImage,
-            // so our SBufferCopy should reflect that
-            asset::IImage::SBufferCopy copyRegion = {};
-            copyRegion.bufferOffset = 0ull;
-            copyRegion.bufferRowLength = cpuimg->getCreationParameters().extent.width;
-            copyRegion.bufferImageHeight = 0u;
-            copyRegion.imageSubresource.aspectMask = asset::IImage::EAF_COLOR_BIT; // Todo(achal): all bits?
-            copyRegion.imageSubresource.mipLevel = 0u;
-            copyRegion.imageSubresource.baseArrayLayer = 0u;
-            copyRegion.imageSubresource.layerCount = cpuimg->getCreationParameters().arrayLayers;
-            copyRegion.imageOffset = { 0u,0u,0u };
-            copyRegion.imageExtent = cpuimg->getCreationParameters().extent;
-
-            cmdbuf_transfer->copyBufferToImage(buf.get(), img, asset::IImage::EL_TRANSFER_DST_OPTIMAL, 1u, &copyRegion);
-        }
-#endif
+        auto regions = cpuimg->getRegions();
+        _params.utilities->updateImageViaStagingBuffer(
+            cmdbuf_transfer.get(), transfer_fence.get(), _params.perQueue[EQU_TRANSFER].queue,
+            cpuimg->getBuffer(), cpuimg->getCreationParameters().format, gpuimg, asset::IImage::EL_TRANSFER_DST_OPTIMAL, regions,
```

Comment:
why do you want the image to be in the `EL_TRANSFER_DST_OPTIMAL` layout afterwards?

Reply:
We need the image to be in `EL_TRANSFER_DST_OPTIMAL` for the image copy.
the lines you have highlighted does not do any layout transitions, we simply need the dstImage's layout for the copy commands inside `updateImageViaStagingBuffer`, further layout transitions are handled here after `cmdUpload` https://github.com/Devsh-Graphics-Programming/Nabla/blob/a344d3e68b81ad3a66f047ff215180db3a816d99/include/nbl/video/utilities/IGPUObjectFromAssetConverter.h#L1091

PS. I simply added format promotion + imageUpload to this function; The whole conversion function and specially the layout transitions here will be revised as part of `AssetConverter 2.0`

Reply:
yeah I got confused because `dstImageLayout` is usually an argument name for something that can perform a layout transition

for `vkCmdCopyBufferToImage` its just some extra constraint param which must match the actual image layout at the time the command executes.

TL;DR the argument should be renamed to `currentDstImageLayout` to stop confusing people using `IUtilities` method